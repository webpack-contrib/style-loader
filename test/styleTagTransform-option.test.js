/* eslint-env browser */

import {
  compile,
  getCompiler,
  getEntryByInjectType,
  getErrors,
  getWarnings,
  runInJsDom,
} from "./helpers/index";

describe('"styleTagTransform" option', () => {
  it(`should work when the "styleTagTransform" option is not specify`, async () => {
    const entry = getEntryByInjectType("simple.js", "styleTag");
    const compiler = getCompiler(entry, {
      injectType: "styleTag",
    });
    const stats = await compile(compiler);

    runInJsDom("main.bundle.js", compiler, stats, (dom) => {
      expect(dom.serialize()).toMatchSnapshot("DOM");
    });

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it(`should work when the "styleTagTransform" option is specify`, async () => {
    const entry = getEntryByInjectType("simple.js", "styleTag");
    const compiler = getCompiler(entry, {
      injectType: "styleTag",
      // eslint-disable-next-line object-shorthand,func-names
      styleTagTransform: function (css, style) {
        // eslint-disable-next-line no-param-reassign
        style.innerHTML = `${css}.modify{}\n`;

        document.head.appendChild(style);
      },
    });
    const stats = await compile(compiler);

    runInJsDom("main.bundle.js", compiler, stats, (dom) => {
      expect(dom.serialize()).toMatchSnapshot("DOM");
    });

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it(`should work when the "styleTagTransform" option is specify and injectType lazyStyleTag`, async () => {
    const entry = getEntryByInjectType("simple.js", "lazyStyleTag");
    const compiler = getCompiler(entry, {
      injectType: "lazyStyleTag",
    });
    const stats = await compile(compiler);

    runInJsDom("main.bundle.js", compiler, stats, (dom) => {
      expect(dom.serialize()).toMatchSnapshot("DOM");
    });

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it(`should work when the "styleTagTransform" option is not specify and injectType lazyStyleTag`, async () => {
    const entry = getEntryByInjectType("simple.js", "lazyStyleTag");
    const compiler = getCompiler(entry, {
      injectType: "lazyStyleTag",
      // eslint-disable-next-line object-shorthand,func-names
      styleTagTransform: function (css, style) {
        // eslint-disable-next-line no-param-reassign
        style.innerHTML = `${css}.modify{}\n`;

        document.head.appendChild(style);
      },
    });
    const stats = await compile(compiler);

    runInJsDom("main.bundle.js", compiler, stats, (dom) => {
      expect(dom.serialize()).toMatchSnapshot("DOM");
    });

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it(`should work when the "styleTagTransform" option is path to module and injectType lazyStyleTag`, async () => {
    const entry = getEntryByInjectType("simple.js", "lazyStyleTag");
    const compiler = getCompiler(entry, {
      injectType: "lazyStyleTag",
      styleTagTransform: require.resolve("./fixtures/styleTagTransform"),
    });
    const stats = await compile(compiler);

    runInJsDom("main.bundle.js", compiler, stats, (dom) => {
      expect(dom.serialize()).toMatchSnapshot("DOM");
    });

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it(`should "styleTagTransform" function path added to buildDependencies when injectType lazyStyleTag`, async () => {
    const styleTagTransformFn = require.resolve("./fixtures/styleTagTransform");
    const entry = getEntryByInjectType("simple.js", "lazyStyleTag");
    const compiler = getCompiler(entry, {
      injectType: "lazyStyleTag",
      styleTagTransform: styleTagTransformFn,
    });
    const stats = await compile(compiler);
    const { buildDependencies } = stats.compilation;

    runInJsDom("main.bundle.js", compiler, stats, (dom) => {
      expect(dom.serialize()).toMatchSnapshot("DOM");
    });

    expect(buildDependencies.has(styleTagTransformFn)).toBe(true);
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
