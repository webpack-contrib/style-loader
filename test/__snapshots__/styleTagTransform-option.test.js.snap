// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"styleTagTransform" option should "styleTagTransform" function path added to buildDependencies when injectType lazyStyleTag: DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>body {
  color: red;
}
.modify{}
</style><style>h1 {
  color: blue;
}
.modify{}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


</body></html>"
`;

exports[`"styleTagTransform" option should "styleTagTransform" function path added to buildDependencies when injectType lazyStyleTag: errors 1`] = `Array []`;

exports[`"styleTagTransform" option should "styleTagTransform" function path added to buildDependencies when injectType lazyStyleTag: warnings 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is not specify and injectType lazyStyleTag: DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>body {
  color: red;
}
.modify{}
</style><style>h1 {
  color: blue;
}
.modify{}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


</body></html>"
`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is not specify and injectType lazyStyleTag: errors 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is not specify and injectType lazyStyleTag: warnings 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is not specify: DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>body {
  color: red;
}
</style><style>h1 {
  color: blue;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


</body></html>"
`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is not specify: errors 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is not specify: warnings 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is path to module and injectType lazyStyleTag: DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>body {
  color: red;
}
.modify{}
</style><style>h1 {
  color: blue;
}
.modify{}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


</body></html>"
`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is path to module and injectType lazyStyleTag: errors 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is path to module and injectType lazyStyleTag: warnings 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is specify and injectType lazyStyleTag: DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>body {
  color: red;
}
</style><style>h1 {
  color: blue;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


</body></html>"
`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is specify and injectType lazyStyleTag: errors 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is specify and injectType lazyStyleTag: warnings 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is specify: DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>body {
  color: red;
}
.modify{}
</style><style>h1 {
  color: blue;
}
.modify{}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


</body></html>"
`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is specify: errors 1`] = `Array []`;

exports[`"styleTagTransform" option should work when the "styleTagTransform" option is specify: warnings 1`] = `Array []`;
